nemo/collections/asr/parts/submodules/rnnt_loop_labels_computer.py

```diff
@@ -245,8 +419,331 @@ def __call__(
                 time_indices += force_blank_mask  # emit blank => advance time indices
                 # update safe_time_indices, non-blocking
                 torch.minimum(time_indices, last_timesteps, out=safe_time_indices)
-                # same as: active_mask = time_indices < out_len
-                torch.less(time_indices, out_len, out=active_mask)
+                # same as: active_mask = time_indices < encoder_output_length
+                torch.less(time_indices, encoder_output_length, out=active_mask)
         if use_alignments:
             return batched_hyps, alignments, last_decoder_state
         return batched_hyps, None, last_decoder_state
+
+    def loop_labels_cuda_graphs(
+        self, encoder_output: torch.Tensor, encoder_output_length: torch.Tensor,
+    ) -> Tuple[rnnt_utils.BatchedHyps, Optional[rnnt_utils.BatchedAlignments], Any]:
+        """
+        Implementation with CUDA graphs.
+
+        Args:
+            encoder_output: output from the encoder
+            encoder_output_length: lengths of the utterances in `encoder_output`
+        """
+        # do not recalculate joint projection, project only once
+        encoder_output = self.joint.project_encoder(encoder_output)
+        current_batch_size = encoder_output.shape[0]
+        current_max_time = encoder_output.shape[1]
+
+        if torch.is_autocast_enabled():
+            encoder_output = encoder_output.to(torch.get_autocast_gpu_dtype())
+
+        # init or reinit graph
+        if self.state is None or self.state.need_reinit(encoder_output):
+            self._graph_reinitialize(encoder_output, encoder_output_length)
+
+        # copy (projected) encoder output and lenghts
+        self.state.encoder_output_projected[:current_batch_size, :current_max_time, ...].copy_(encoder_output)
```

Comment:
Just a note for the future. If we get torch.while_loop to work with conditional nodes, we would be able to speed up these three torch ops by running them in three parallel cuda streams (I think this is called vertical or horizontal fusion), assuming torch inductor or thunder can do that right now.